File: permutations.py
import sys

def permutation(input_, start_):
    l = len(input_)
    
    if start_ >= l:
        print(input_)
    else:
        permutation(input_, start_ + 1)
    
    for i in range(start_ + 1, l):
        newinput_ = swap(input_, start_, i)
        permutation(newinput_, start_ + 1)

def swap(s, i, j):
    t = list(s)
    t[i], t[j] = t[j], t[i]
    return ''.join(t)

def main():
    if len(sys.argv) < 2:
        sys.exit("not correct usage")
    input_ = sys.argv[1]
    permutation(input_, 0)

if __name__ == "__main__":
    main()

File: permutations.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

void permutation(string input, int start) {
    int l = input.length();
    
    if (start >= l){
        cout << input << "\n";
    } else {
        permutation(input, start + 1);
    }
    
    for (int i = start + 1; i < l; i++){
	string newinput(input);
        swap(newinput[start], newinput[i]);
        permutation(newinput, start + 1);
    }
}

int main(){
	string input;
	getline(std::cin, input);
	permutation(input, 0);
}
